db/commitlog/commitlog.cc

```diff
@@ -1344,18 +1451,195 @@ db::commitlog::segment_manager::segment_manager(config c)
+
+    auto writes = handle_writes();
+    auto replenish = replenish_reserve();
+    auto shutdown_progress = make_ready_future<>();
+    auto last_sync = message_clock::now();
+    auto sync_period = std::chrono::milliseconds(cfg.commitlog_sync_period_in_ms);
+
+    while (_state != state::terminated || !_message_queue.empty()) {
+        auto now = message_clock::now();
+        auto next_wakeup = now + sync_period;
+
+        if (cfg.mode == sync_mode::PERIODIC) {
+            if (!_segments.empty() && _segments.back()->is_still_allocating()) {
+                last_sync = std::min(last_sync, _segments.back()->sync_time());
+            }
+            next_wakeup = std::min(next_wakeup, last_sync + sync_period);
+        }
+
+        // handling timeout exception with a "handle_exception" and just assuming
+        // it is this error, and not an actual one, since we don't sent exceptions
+        // through the cond/queue.
+        // handling it in a handle_exception handler prevents actual throw from
+        // occuring, which prevents unwarranted exception stalls here every N ms
+        if (_message_queue.empty()) {
+            try {
+                co_await _message_queue.wait(next_wakeup);
+            } catch (condition_variable_timed_out&) {
+            }
+        }
+
+        while (!_message_queue.empty()) {
+            auto m = _message_queue.pop();
+
+            co_await std::visit(overloaded_functor {
+                [&](const shutdown_request&) { 
+                    shutdown = true;
+                    return make_ready_future<>(); 
+                },
+                [&](const close_request& r) {
+                    return r.segment->perform_close_request(r);
+                },
+            }, m);
+        }
+
+        if (cfg.mode == sync_mode::PERIODIC) { 
+            now = message_clock::now();
+            if (now >= next_wakeup || (now - last_sync) >= sync_period) {
+                do_periodic_syncs();
+                last_sync = now;
+            }
+        }
+
+        do_periodic_flush_callbacks();
+        discard_unused_segments();
+        co_await do_pending_deletes();
+
+        if (shutdown && shutdown_progress.available()) {
+            shutdown_progress = handle_shutdown_stage();
+        }
+    }
+
+    // since tasks we drained might have had segments kept alive,
+    // we need to do a final sweep here.
+    discard_unused_segments();
+
+    _write_queue.close();
+    co_await std::move(writes);
+
+    co_await clear_reserve_segments();
+    co_await std::move(replenish);
+    co_await do_pending_deletes();    
+}
+
+future<> db::commitlog::segment_manager::handle_writes() noexcept {
+    while (!_write_queue.closed() || !_write_queue.empty()) {
+        co_await _write_queue.wait();
+
+        while (!_write_queue.empty()) {
+            auto wr = _write_queue.pop();
+            auto s = wr.segment;
+            try {
+                co_await s->perform_write_request(wr);
```

Comment:
Yes, one can parallelize, but then we are back to either chained continuations or storing futures in a vector (because we want to avoid detached futures). In general, I would prefer (if it shows it could help) to do write coalescing first - i.e. merge N write requests into 1 (if multiple messages queued).